pkg/registries/registries.go

```diff
@@ -71,24 +71,29 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
 	}
 	sort.Strings(sources)
 	// Convert the sets of mirrors
-	res := []apioperatorsv1alpha1.RepositoryDigestMirrors{}
+	res := []apicfgv1.RepositoryDigestMirrors{}
 	for _, source := range sources {
+		allowMirrorByTagsFlag := false
 		ds := disjointSets[source]
 		topoGraph := newTopoGraph()
 		for _, set := range *ds {
+			// AllowMirrorByTags is only set once for each source in imagecontentpolicies.config.openshift.io
```

Comment:
The external [API](https://github.com/openshift/api/blob/master/config/v1/types_image_content_policy.go#L46) added the kubebuilder validation to use source as key of the map since I saw the mirrors are merged in the mergedMirrorSets() use the source as the key of the map. We can change that API since it's a new CRD we want to add.
If enables setting multiple same source, how should we handle the conflict `allowMirrorByTags` for the same source? e.g. allowMirrorByTags is `true` for source:A, mirrors:[mirror0, mirror1, mirror2], and `false` for source:A, mirrors: [mirror1, mirror2]. Do we still use current logic to merge the mirrors by the source? @mtrmac WDYT?

Reply:
> The external [API](https://github.com/openshift/api/blob/master/config/v1/types_image_content_policy.go#L46) added the kubebuilder validation to use source as key of the map since I saw the mirrors are merged in the mergedMirrorSets() use the source as the key of the map. We can change that API since it's a new CRD we want to add.

I can’t find a precise definition of the `listType=map` annotation, but if it means that the data should be expressible as a `map[source]…`, with each `source` value being used at most once, that looks, from a historical point of view at least, incorrect, or at least insufficient.

(If we wanted to enforce uniqueness, the code in this subpackage could have just refused to work on duplicates, instead of doing all of the sorting and merging that it currently does.)

On the contrary, https://github.com/openshift/api/blob/636513e97fdaf2dc65efd95e8c239f892057f791/config/v1/types_image_content_policy.go#L39-L43 _explicitly_ documents that multiple entries with the same “source” is a valid configuration, and prescribes the kind of merging this package does.

(This means that an administrator or an operator can always add an ICSP CR object, without manually merging with any existing data, and leaving the merging work to be done by MCO or other consumers.)

As yet another argument, even if the `ICP.Spec.RepositoryDigestMirrors` field were defined as a map with unique `source` values, AFAICS MCO is listing _all_ (currently) ICSP resources and merges the data (note that the input of `mergedMirrorSets` is an _array_ of `[]*apicfgv1.ImageContentPolicy`; so unique `source` values within one CR doesn’t mean anything WRT uniqueness across CRs.

So if we can still change the CRD, I’d prefer the map-like behavior annotation to be dropped (if that’s what the annotation is).

---

Actually, if we can still change the CRD, shouldn’t `RepositoryDigestMirrors` be just `RepositoryMirrors`, now that they are not always digest-restricted?

---

> If enables setting multiple same source, how should we handle the conflict `allowMirrorByTags` for the same source? e.g. allowMirrorByTags is `true` for source:A, mirrors:[mirror0, mirror1, mirror2], and `false` for source:A, mirrors: [mirror1, mirror2]. Do we still use current logic to merge the mirrors by the source? @mtrmac WDYT?

Yes, that’s the problem I was asking about.

We can’t just _not merge_ the entries; `registries.conf` only uses one value per a prefix, so creating a config file that has two entries for the same `source` would just not represent some of the mirrors.

The technically correct/clean implementation, given the current CRD, would be to reject configurations that specify conflicting values; that either does exactly what the user asked, or refuses to do anything.

OTOH that also means that the “just add an ICSP CR” use case described above no longer works that easily, because the newly added CR would need to have an `AllowMirrorByTags` value that matches any pre-existing CRs.

Hence I don’t have a strong opinion.

It’s generally easier to be more restrictive and loosen restrictions, than to ship a less restrictive implementation and tighten it later (possibly breaking users), so I’m leaning towards the stricter approach of rejecting conflicts, but that’s not a strong argument and there could well be a good reason not to do that.

Reply:
> On the contrary, https://github.com/openshift/api/blob/636513e97fdaf2dc65efd95e8c239f892057f791/config/v1/types_image_content_policy.go#L39-L43 explicitly documents that multiple entries with the same “source” is a valid configuration, and prescribes the kind of merging this package does. 

The `multiple entries with the same “source” is a valid` was a mistake. The listType=map will limit the same `source` can only be used once. I don't think the openshift/api reviewers will disallow the changes on this CRD since it was just newly created for the implementation of allowMirrorByTags. I will create a new PR and the comment can leave it as is if the listType=map is removed.

> Actually, if we can still change the CRD, shouldn’t RepositoryDigestMirrors be just RepositoryMirrors, now that they are not always digest-restricted?

AllowMirrorByTags makes the mirror-by-tags only configurable but does not change the current default setting. The digest restriction still exists as a default. Should we just keep the Digest in the name RepositoryDigestMirrors?

> The technically correct/clean implementation, given the current CRD, would be to reject configurations that specify conflicting values; that either does exactly what the user asked, or refuses to do anything.
> 
> OTOH that also means that the “just add an ICSP CR” use case described above no longer works that easily, because the newly added CR would need to have an AllowMirrorByTags value that matches any pre-existing CRs.

I am confused about the definitions of conflict configurations. So the rule of rejecting conflicting values would be, going through the list of  []*apicfgv1.ImageContentPolicy, if `allowMirrorByTags` value is different for the same source, we think of it as conflict?  Although the user may just want to update the `allowMirrorByTags`, we have no idea which value is the user expected.
Yes, the “just add an ICSP CR” use case is not that easy. When the `allowMirrorByTags` value needs to change for the source that exists in pre-exising CR, the existed CRs should be deleted before applying a new one in order to avoid the conflict.

Reply:
> and not the comment?

… saying that multiple entries with the same “source” is valid.

Reply:
> I’m a bit confused, just to confirm; do I understand it correctly that you are proposing to remove the `listType=map` and not the comment?

@mtrmac  Yes. Keep the comment about `multiple same source`, and remove the `listType=map` validation.